File: scripts/service_fixtures.py
# SPDX-License-Identifier: Apache-2.0
import json
from copy import deepcopy
from pathlib import Path
from psp_cdl_api_server import SecurityService
SUITE=json.loads((Path(__file__).resolve().parents[1]/'conformance/vectors/services/profile-0.1.json').read_text(encoding='utf-8'))
def fixture(case=None):
    case=case or {}
    class Host:
        def __init__(self):
            self.snapshot=deepcopy(SUITE['snapshot'])
            self.resolutions=0
            self.token='public-token-a'
            for key in self.snapshot['verification']['keys']:
                key['material']=bytes.fromhex(key.pop('materialHex'))
                if case.get('revoked'): key['status']='revoked'
            if case.get('allow'): self.snapshot['resources'][0]['capabilities']=[]
        def authenticate(self,token): return SUITE['principals'].get(token)
        def resolve(self,principal,operation_id):
            self.resolutions+=1
            if case.get('backendFailure'): raise RuntimeError('PRIVATE_BACKEND_DETAIL')
            return self.snapshot
        def now(self): return case.get('now',SUITE['now'])
    host=Host()
    headers=[] if case.get('token','public-token-a') is None else [['Authorization','Bearer '+case.get('token','public-token-a')]]
    headers+=[['Content-Type',case.get('contentType','application/json')]]+case.get('extraHeaders',[])
    request={'method':case.get('method','POST'),'path':case.get('path','/v1/policy/evaluate'),'headers':headers,'body':case.get('rawBody',json.dumps(case.get('request',{'operation_id':'op-1'}))).encode('utf-8')}
    return SecurityService(host),request,host

File: scripts/service-fixtures.mjs
// SPDX-License-Identifier: Apache-2.0
import {readFileSync} from 'node:fs';
import {SecurityService} from '@psp-cdl/api-server';
export const suite=JSON.parse(readFileSync(new URL('../conformance/vectors/services/profile-0.1.json',import.meta.url),'utf8'));
export function fixture(c={}) {
  const snapshot=structuredClone(suite.snapshot);
  snapshot.verification.keys=snapshot.verification.keys.map(({materialHex,...key})=>({...key,material:Buffer.from(materialHex,'hex')}));
  if(c.allow) snapshot.resources[0].capabilities=[];
  if(c.revoked) snapshot.verification.keys[0].status='revoked';
  let resolutions=0, token='public-token-a';
  const service=new SecurityService({authenticate:t=>Object.hasOwn(suite.principals,t)?suite.principals[t]:null,resolve:()=>{resolutions++;if(c.backendFailure)throw new Error('PRIVATE_BACKEND_DETAIL');return snapshot;},now:()=>c.now??suite.now});
  const request={method:c.method??'POST',path:c.path??'/v1/policy/evaluate',headers:[...(c.token===null?[]:[['Authorization','Bearer '+(c.token??'public-token-a')]]),['Content-Type',c.contentType??'application/json'],...(c.extraHeaders??[])],body:Buffer.from(c.rawBody??JSON.stringify(c.request??{operation_id:'op-1'}))};
  return {service,request,snapshot,resolutions:()=>resolutions,credential:()=>token,setToken:t=>{token=t;}};
}
